### PR TITLE
change GetJSON to accept generic predicate instead of glob pattern

### PR DESCRIPTION
**Pull request by pnemtsev**

### CODE REVIEW

File: meter.go [hunk 1]
```diff
@@ -270,8 +260,8 @@ func Get(counterName string) Counter {
 
 // GetJSON filters counters by glob pattern and returns them as a json marshaled
 // map.
-func GetJSON(pattern string) ([]byte, error) {
-	return getJSON(Default, pattern)
+func GetJSON(predicate func(string) bool) ([]byte, error) {
```
Comment (dshil) [inline meter.go]: We've stopped using the glob by default, so the comment should be reworked.

File: meter_test.go [hunk 2]
```diff
@@ -300,14 +321,11 @@ func TestMetricsGetJSONGlobPatterns(t *testing.T) {
 	require.Nil(t, metrics.Register("aaac", counter4))
 
 	for _, tCase := range tests {
-		b, err := metrics.GetJSON(tCase.pattern)
+		g := glob.MustCompile(tCase.pattern)
```
Comment (dshil) [inline meter_test.go]: Could you please separate the glob related tests from the general `meter` tests?
I'd suggest to put them under `_examples/glob_test.go`.

File: meter_test.go [hunk 3]
```diff
@@ -220,11 +221,21 @@ func TestMetricsGetJSON(t *testing.T) {
 	counter2.Set(42)
 	require.Nil(t, metrics.Register("counter2", counter2))
 
-	b, err := metrics.GetJSON("count*")
+	b, err := metrics.GetJSON(func(key string) bool {
```
Comment (dshil) [inline meter_test.go]: I'd suggest to add a test where zero counters are suitable for our pattern.

File: meter.go [hunk 4]
```diff
@@ -108,31 +107,22 @@ func getCounter(m *Metrics, counterName string) Counter {
...
 			result[k] = v.Get()
 		}
 	}
 
-	b, err := json.Marshal(result)
-	if err != nil {
-		return nil, err
-	}
-	return b, nil
+	return json.Marshal(result)
```
Comment (dshil) [inline meter.go]: Instead of `json.Marshal` I'd suggest to use:

```go
package gometer

import (
	"bytes"
	"fmt"
)

type jsonFormatter struct{}

func (j *jsonFormatter) Format(counters map[string]Counter) []byte {
	var buf bytes.Buffer
	countersNum := len(counters)

	panicIfErr(buf.WriteByte('{'))

	for i, n := range sortedKeys(counters) {
		_, err := fmt.Fprintf(&buf, `"%v":%v`, n, counters[n].Get())
		panicIfErr(err)

		if i != countersNum-1 {
			panicIfErr(buf.WriteByte(','))
		}
	}

	panicIfErr(buf.WriteByte('}'))

	return buf.Bytes()
}

func panicIfErr(err error) {
	if err != nil {
		panic(err)
	}
}
```

`sortedKeys` can be found in `formatter.go`

As a result we'll have sorted metrics and reduce and number of allocation, that sounds good if there are tons of metrics.

File: meter_test.go (hunk 3 above)
Comment (pnemtsev) [reply meter_test.go]: I've added tests for an empty pattern.

File: _examples/glob_test.go [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+			pattern:  "*b*",
+			expected: `{"abc": 10, "abb": 42}`,
+		},
+		{
+			pattern:  "??[ab]*",
+			expected: `{"abb": 42, "aaac":17}`,
+		},
+	}
+
+	counter1 := new(gometer.DefaultCounter)
```
Comment (dshil) [inline _examples/glob_test.go]: We can simplify it a bit, e.g.:

```go
	for k, v := range map[string]int64{
		"aaab": 2,
	} {
		c := &gometer.DefaultCounter{}
		c.Set(v)
		if err := metrics.Register(k, &c); err != nil {
			fmt.Println(err)
			return
		}
	}
```

File: meter.go [hunk 6]
```diff
@@ -106,33 +104,25 @@ func getCounter(m *Metrics, counterName string) Counter {
...
-func getJSON(m *Metrics, pattern string) ([]byte, error) {
-	g, err := glob.Compile(pattern)
-	if err != nil {
-		return nil, err
-	}
-
-	result := make(map[string]int64)
+func getJSON(m *Metrics, predicate func(string) bool) []byte {
+	result := make(map[string]Counter)
+	formatter := jsonFormatter{}
```
Comment (dshil) [inline meter.go]: Moving locking/unlocking from the top to some random place can cause a simple forgetting to lock or unlock. To avoid such cases and to reduce a WTF level let's use the following:

```go

func (f *foo) bar() {
f.mu.Lock()
defer f.mu.Unlock()

// other stuff
}
```

**Comment by dshil** [comment]
LGTM
